src/bin/notification-proxy-client.rs

```diff
@@ -309,6 +314,15 @@ async fn client_server() {
                 .map
                 .remove(&sequence)
                 .expect("server violated the protocol")
+                // marmarek: is sending real ID back to the client a good idea? the specification
+                // mandates ID to be unique (until wraps) for the notification server uptime, but
+                // if the GUI domain's server is restarted but this one isn't, it may result in a
+                // duplicated ID returned (much) earlier.
+                // Maybe better send sequence number as the ID instead and keep mapping on the
+                // server side? and remove stale entries when notification is closed (you get a
+                // dbus signal) or when the actual notification server gets restart (name owner
+                // change signal).
```

Comment:
I still need to handle name changes.

Reply:
I couldn't check it as the current version doesn't build, but how does it handle GUI in dom0 restart (user logout and login for example)? It may mean that session dbus daemon will be restarted, not just the notification daemon. In that case, the notification-proxy-daemon should IMO restart too (so, the qrexec service should terminate and then the connection should be re-established), to fetch new session environment etc (`DBUS_SESSION_BUS_ADDRESS` may change, theoretically it doesn't need to be always the same). But then the agent need to handle service termination and re-establish the connection.

My point is - is the separate protocol message necessary for handling notification daemon restart? It feels like it handles only less common case (restarting notification daemon while the rest of the DE remains running) but doesn't handle the case of whole DE restart. Or maybe it does handle that case too?

As for re-establishing qrexec connection, it can be added later, as it may be a mit more complicated. Specifically, it should have some rate limits, handling request denied or so. Basically to not spin in a loop when no user is logged in.